OpenSim/Common/DataAdapter.h

```diff
@@ -0,0 +1,225 @@
+ * not use this file except in compliance with the License. You may obtain a  *
+ * copy of the License at http://www.apache.org/licenses/LICENSE-2.0.         *
+ *                                                                            *
+ * Unless required by applicable law or agreed to in writing, software        *
+ * distributed under the License is distributed on an "AS IS" BASIS,          *
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.   *
+ * See the License for the specific language governing permissions and        *
+ * limitations under the License.                                             *
+ * -------------------------------------------------------------------------- */
+
+/** @file
+* This file defines the abstract DataAdapter class, which is used to access
+* Data sources and populate DataTables for consumption in OpenSim.
+*/
+
+#include "osimCommonDLL.h"
+#include "Exception.h"
+#include <SimTKcommon.h>
+#include <string>
+#include <memory>
+#include <ios>
+#include <map>
+
+namespace OpenSim {
+
+class AbstractDataTable;
+//=============================================================================
+//=============================================================================
+/**
+ * DataAdapter is an abstract class defining an interface for reading/writing
+ * in/out the contents of a DataTable. It enables access to/from various data
+ * sources/sinks such as: streams, files, databases and devices. The DataTable
+ * is independent of the form and format of the data in/out of the source/sink.
+ * Concrete classes handle the details (e.g. format, sequential access, etc...) 
+ * associated with a particular data source/sink.
+ *
+ * The base DataAdapter contains a static registry to serve as a factory for 
+ * concrete DataAdpaters, given a string identifier of the type of adapter.
+ * The adapter knows the source format and data flow (read, write, both).
+ * String identifiers can be associated with file formats according to known
+ * file extensions.
+ *
+ * @author Ajay Seth
+ */
+class OSIMCOMMON_API DataAdapter { 
+//=============================================================================
+// METHODS
+//=============================================================================
+public:
+    virtual ~DataAdapter(){};
+
+    virtual DataAdapter* clone() const = 0;
+
+    /** Creator of concrete DataAdapter(s) for the specified source type by its
+        unique identifier (string). For example, for file based sources, a 
+        component can acquire the necessary adapter instance to read the data
+        by the file extension if the extension is used as its identifier.
+    @param[in] identifier string used to uniquely identify the concrete adapter
+               that processes data from/to sources/sinks of a particular type.
+               For file adapters, .*/
+    static DataAdapter* createAdapter(const std::string& identifier);
```

Comment:
More questions, are these case sensitive? Also are we restricted to one Adapter per extension or can we replace Adapters (e.g. c3d files changed contents/format over time)